### ISSUE TITLE
Ada 2012-like Dynamic_Predicate support for types

### ISSUE THREAD

**Issue by wizzardx** [nim-lang/Nim#7122]
Hi.

Nim already has range checked integers and also sub-ranges, distinct types, etc.

A lot of this seems to be borrowed from Ada, which is a huge plus in my opinion.

Anyhow, one of Ada's really nice type features is the ability to define "Dynamic_Predicate". That is, rules that must always be followed by a given variable.

Here's an example of that, an example of Ada 2012 subtype predicates (copypasted from somewhere on the web):

```
-- SSN format: ###-##-####
Subtype Social_Security_Number is String(1..11)
  with Dynamic_Predicate =>
    (for all Index in Social_Security_Number'Range =>
      (case Index is
       when 4|7 => Social_Security_Number(Index) = '-',
       when others => Social_Security_Number(Index) in '0'..'9'
      )
     );
```

Another feature from Ada that this can help support is allowing ranged types that cover floating point numbers ranges, and not just integers, for instance.

It's a lot like "dependent types", except implemented at runtime rather than compile time.

**Comment by andreaferretti** [comment]
If this works at runtime, it can be implemented in a library. It looks to me very much like [contracts](https://github.com/Udiknedormin/NimContracts), would that work for you?

**Comment by wizzardx** [comment]
Not entirely the same.

There's a good overview over here:

https://en.wikibooks.org/wiki/Ada_Programming/Contract_Based_Programming

Contracts are for a block of code, predicates follow the types around wherever they get used.

It's similar to eg, how nim does debug checks for integers going out of range or overflowing at runtime (eg: the nim "Natural" numeric type), freeing you from needing to do asserts for that on each function. Also, it helps to convey the meaning of the type more clearly.

**Comment by andreaferretti** [comment]
I feel like attaching this runtime check to the type would repeat the same check many times redundantly

**Comment by wizzardx** [comment]
Only when variables get created or modified. But not when say just passed around as arguments to functions.

(this would be eg, one big point in it's favor, vs eg having contract assertions in each function that requires that a sequence passed into it has between 2 and 10 items, and must be sorted).

More or less the implementation would be similar to how nim Range types work, when compiler isn't able to statically infer an invalid assignment or update.

Also, like the other runtime checks, it only happens in debug mode.

A more indirect way would be eg, giving each type it's own module, and only exposing some operators with "*", which have assert lines in them. You could do the same to implement Natural too, or range-limited integers too, for instance.

Basically, it's a lightweight version of having classes with class invariants (Eiffel-style contract dev, which eg D lang uses, but Ada doesn't use).

**Comment by krux02** [comment]
Have you tried implementing this feature with macros already?

**Comment by wizzardx** [comment]
Making a macro would mean knowing how to do it in regular verbose Nim.

eg, let's say that Nim didn't have a built-in subrange type, but I wanted to ensure that an integer stays between 2 and 5.

With regular verbose nim, I don't know how to tell where or if that integer var gets updated to an invalid value:

```
var a = 5

foo(a)
bar(a)
qux(a)
```

I'd have to do it something like this in a macro block:

```
var a = 5
assert a >= 2 and a <= 5

foo(a)
assert a >= 2 and a <= 5

bar(a)
assert a >= 2 and a <= 5

qux(a)
assert a >= 2 and a <= 5
```

And then also putting "assert a >= 2 and a <= 5" anywhere else in other calls outside of the macro block, where there's some chance that "a" might be modified.

Yet, Nim with the Subrange type, is able to tell directly where a is modified, and then just put the subrange check there, and not next to every operator where a is passed along as a reference var that might or might not be modified.

I think it must work by some kind of "magic function" details in the compiler, eg Nim knows that specific primitive functions are places where range checks need to be added in debug mode (=, inc, etc).

What I'm proposing here would be more or less, being able to do something like adding extra customizable logic to where nim would normally add it's range/out of bounds/etc debug checks.

**Comment by krux02** [comment]
well I think you can get pretty far with converter fucntions and a distinct type:

```Nim
type
  SocialSecurityNumber = object
    value*: string

converter toSocialSecurityNumber(arg: string): SocialSecurityNumber =
  for i, c in arg:
    case i
    of 4 | 7:
      assert(arg[i] == '-')
    else:
      assert(arg[i].isDigit)

  result.value = arg

converter toString(arg: SocialSecurityNumber): string =
    result = arg.value
```

**Comment by wizzardx** [comment]
Very true :-)

I've used and recommended that approach previously, too.

Mainly I've drunk the Ada cool aid recently and want more of that in Nim :-)

**Comment by Araq** [comment]
As nice as this feature would be, we're focussing on bugfixes and features that pull in their weight. We are not looking to open another can of worms.
